crypto/defaults.c

```diff
@@ -0,0 +1,200 @@
+ */
+static char openssldir[MAX_PATH + 1];
+
+/**
+ * @brief The pointer to the opennsldir buffer
+ */
+static char *openssldirptr = NULL;
+
+/**
+ * @brief The directory where OpenSSL engines are located.
+ */
+
+static char enginesdir[MAX_PATH + 1];
+
+/**
+ * @brief The pointer to the enginesdir buffer
+ */
+static char *enginesdirptr = NULL;
+
+/**
+ * @brief The directory where OpenSSL modules are located.
+ */
+static char modulesdir[MAX_PATH + 1];
+
+/**
+ * @brief The pointer to the modulesdir buffer
+ */
+static char *modulesdirptr = NULL;
+
+/**
+ * @brief Get the list of Windows registry directories.
+ *
+ * This function retrieves a list of Windows registry directories.
+ *
+ * @return A pointer to a char array containing the registry directories.
+ */
+static char *get_windows_regdirs(char *dst, LPCTSTR valuename)
+{
+    DWORD keysize;
+    DWORD ktype;
+    HKEY hkey;
+    LSTATUS ret;
+    DWORD index = 0;
+    LPCTCH tempstr = NULL;
+    char *retval = NULL;
+
+    ret = RegOpenKeyEx(HKEY_LOCAL_MACHINE,
+                       TEXT(REGISTRY_KEY), KEY_WOW64_32KEY,
+                       KEY_QUERY_VALUE, &hkey);
+    if (ret != ERROR_SUCCESS)
+        goto out;
+
+    ret = RegQueryValueEx(hkey, valuename, NULL, &ktype, NULL,
+                          &keysize);
+    if (ret != ERROR_SUCCESS)
+        goto out;
+    if (ktype != REG_EXPAND_SZ)
+        goto out;
+    if (keysize > MAX_PATH)
+        goto out;
+
+    keysize++;
+    tempstr = OPENSSL_zalloc(keysize * sizeof(TCHAR));
+
+    if (tempstr == NULL)
+        goto out;
+
+    if (RegQueryValueEx(hkey, valuename,
+                        NULL, &ktype, tempstr, &keysize) != ERROR_SUCCESS)
+        goto out;
+
+    if (!WideCharToMultiByte(CP_UTF8, 0, tempstr, -1, dst, keysize,
+                             NULL, NULL)) 
+        goto out;
+
+    retval = dst;
+out:
+    OPENSSL_free(tempstr);
+    RegCloseKey(hkey);
+    return retval;
+}
+
+static CRYPTO_ONCE defaults_setup_init = CRYPTO_ONCE_STATIC_INIT;
+
+/**
+ * @brief Function to setup default values to run once.
+ * Only used in Windows environments.  Does run time initalization
+ * of openssldir/modulesdir/enginesdir from the registry
+ */
+DEFINE_RUN_ONCE_STATIC(do_defaults_setup)
+{
+    get_windows_regdirs(openssldir, TEXT("OPENSSLDIR"));
+    get_windows_regdirs(enginesdir, TEXT("ENGINESDIR"));
+    get_windows_regdirs(modulesdir, TEXT("MODULESDIR"));
+
+    /*
+     * Set our pointers only if the directories are fetched properly
+     */
+    if (strlen(openssldir))
+        openssldirptr = openssldir;
+
+    if (strlen(enginesdir))
+        enginesdirptr = enginesdir;
+
+    if (strlen(modulesdir))
+        modulesdirptr = modulesdir;
+
+    return 1;
+}
+#endif
+
+/**
+ * @brief Get the directory where OpenSSL is installed.
+ *
+ * @return A pointer to a string containing the OpenSSL directory path.
+ */
+const char *ossl_get_openssldir(void)
+{
+#if defined(_WIN32)
```

Comment:
This changes the semantics also when OSSL_WINCTX is not defined. I thought we do not want to do this breaking change yet. I.e. do:
```suggestion
#if defined(_WIN32) && defined(OSSL_WINCTX)
```

here and in the other functions below.